xcm/xcm-emulator/example/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+	use sp_runtime::traits::AccountIdConversion;
+	use xcm::{v3::prelude::*, VersionedMultiLocation, VersionedXcm};
+	use xcm_emulator::TestExt;
+
+	#[test]
+	fn dmp() {
+		Network::reset();
+
+		let remark = test_runtime::RuntimeCall::System(
+			frame_system::Call::<test_runtime::Runtime>::remark_with_event {
+				remark: "Hello from Kusama!".as_bytes().to_vec(),
+			},
+		);
+		KusamaNet::execute_with(|| {
+			assert_ok!(kusama_runtime::XcmPallet::force_default_xcm_version(
+				kusama_runtime::RuntimeOrigin::root(),
+				Some(3)
+			));
+			assert_ok!(kusama_runtime::XcmPallet::send_xcm(
+				Here,
+				Parachain(1),
+				Xcm(vec![Transact {
+					origin_kind: OriginKind::SovereignAccount,
+					require_weight_at_most: Weight::from_parts(INITIAL_BALANCE as u64, 1024 * 1024),
+					call: remark.encode().into(),
+				}]),
+			));
+		});
+
+		ParachainA::execute_with(|| {
+			use test_runtime::{RuntimeEvent, System};
+			System::events().iter().for_each(|r| println!(">>> {:?}", r.event));
+
+			assert!(System::events().iter().any(|r| matches!(
+				r.event,
+				RuntimeEvent::System(frame_system::Event::Remarked { sender: _, hash: _ })
+			)));
+		});
+	}
+
+	#[test]
+	fn ump() {
+		Network::reset();
+
+		KusamaNet::execute_with(|| {
+			assert_ok!(kusama_runtime::XcmPallet::force_default_xcm_version(
+				kusama_runtime::RuntimeOrigin::root(),
+				Some(3)
+			));
+			let _ = kusama_runtime::Balances::deposit_creating(
+				&ParaId::from(1).into_account_truncating(),
+				1_000_000_000_000,
+			);
+		});
+
+		let remark = kusama_runtime::RuntimeCall::System(frame_system::Call::<
+			kusama_runtime::Runtime,
+		>::remark_with_event {
+			remark: "Hello from Pumpkin!".as_bytes().to_vec(),
+		});
+		ParachainA::execute_with(|| {
+			assert_ok!(test_runtime::PolkadotXcm::force_default_xcm_version(
+				test_runtime::RuntimeOrigin::root(),
+				Some(3)
+			));
+			assert_ok!(test_runtime::PolkadotXcm::send_xcm(
+				Here,
+				Parent,
+				Xcm(vec![
+					UnpaidExecution { weight_limit: Unlimited, check_origin: None },
+					Transact {
+						origin_kind: OriginKind::SovereignAccount,
+						require_weight_at_most: Weight::from_parts(
+							INITIAL_BALANCE as u64,
+							1024 * 1024
+						),
+						call: remark.encode().into(),
+					}
+				]),
+			));
+		});
+
+		KusamaNet::execute_with(|| {
+			use kusama_runtime::{RuntimeEvent, System};
+			// TODO: https://github.com/paritytech/polkadot/pull/6824 or change this call to
+			// force_create_assets like we do in cumulus integration tests.
+			// assert!(System::events().iter().any(|r| matches!(
+			// 	r.event,
+			// 	RuntimeEvent::System(frame_system::Event::Remarked { sender: _, hash: _ })
+			// )));
+			assert!(System::events().iter().any(|r| matches!(
+				r.event,
+				RuntimeEvent::Ump(polkadot_runtime_parachains::ump::Event::ExecutedUpward(
+					_,
+					Outcome::Incomplete(_, XcmError::NoPermission)
+				))
+			)));
+		});
+	}
+
+	#[test]
+	fn xcmp() {
```

Comment:
Do these tests care about the transport? 
```suggestion
	fn para_to_para() {
```

Reply:
The XCM-Emulator is using XCMP/UMP sinks to pass messages around, so it's relevant. Another nice thing about having those names is that they correspond to those in `parachains-integration-tests` repo and potentially to the future parachain tests.

Reply:
These tests should work the same way when HRMP exists. I think Joe's right on this one - it's not that particular.